add special processing of `multigraph_input` upon graph init

dschult opened the pull request:
Fixes: #4720

Adding a keyword argument `multigraph_input=True`
to a graph construction call should treat any incoming input data for
the graph as a dict-of-dict-of-dict multigraph data structure.

This matches the behavior of `nx.to_networkx_graph` which is called by the `__init__` dunder.

Previously, the multigraph_input argument would be added to the graph
attribute dict and ignored when processing the input data.

MridulS commented:
Shouldn't we also add `multigraph_input=True` explicitly to the `__init__` constructors of `Multi(Di)Graph`? 
This wouldn't fix #4824 as we would still need to pass `multigraph_input=True` while giving `MultiGraph` a dict-o-dict-o-dict-o-dict data structure.

```
diff --git a/networkx/classes/multigraph.py b/networkx/classes/multigraph.py
index 1779a616..e3758531 100644
--- a/networkx/classes/multigraph.py
+++ b/networkx/classes/multigraph.py
@@ -306,7 +306,7 @@ class MultiGraph(Graph):

         """
         self.edge_key_dict_factory = self.edge_key_dict_factory
-        Graph.__init__(self, incoming_graph_data, **attr)
+        Graph.__init__(self, incoming_graph_data, multigraph_input=True, **attr)

     @property
     def adj(self):
```

With this PR:

```python
In [1]: import networkx
   ...: g = networkx.MultiGraph({'a': {'b': {0: {'w': 200}, 1: {'w': 201}}}})
   ...: print(g.edges)
   ...: h = networkx.MultiGraph()
   ...: h.add_edges_from([('a', 'b', {"w": 200}), ('a', 'b', {"w": 201})])
   ...: print(h.edges)
[('a', 'b', 0)]
[('a', 'b', 0), ('a', 'b', 1)]
```

After forcing all the incoming data to be `multigraph_input=True`
``` python
In [1]: import networkx
   ...: g = networkx.MultiGraph({'a': {'b': {0: {'w': 200}, 1: {'w': 201}}}})
   ...: print(g.edges)
   ...: h = networkx.MultiGraph()
   ...: h.add_edges_from([('a', 'b', {"w": 200}), ('a', 'b', {"w": 201})])
   ...: print(h.edges)
[('a', 'b', 0), ('a', 'b', 1)]
[('a', 'b', 0), ('a', 'b', 1)]

```

dschult commented:
It's tricky to know what the default should be because the dict-of-dict-of-dict might represent a multigraph, but it might not. So, we shouldn't make the behavior rule out the possibility of someone using multigraph_input=False even while they are making a multigraph.  

Note also that this usage must be extremely rare because no one has reported it in the ~ 10-15 years that this behavior has been in its current form. It's a little strange that we got 2 reports of it within a month :)  Perhaps people are using dict-of-dict-of-dict-of-dict more often than they used to.  It's definitely worth figuring out what the best approach to the API is even if the usage is rare...

At the moment there are two ways to construct a MultiGraph from a dict-of-dict-of-dict-of-dict. The first is `nx.to_network_graph` and the second is with `nx.MultiGraph`. The current code for `to_networkx_graph` provides a flag input argument to allow you to say whether the input data structure is representing a multigraph or not. The current code for `MultiGraph` does not allow that, so the input is treated as a graph (without multiedges) always when using the `MultiGraph` idiom. 

The fix in this PR allows the user to pass a flag input argument to the `MultiGraph` constructor to allow treatment as a multigraph at the cost of **not** allowing a graph attribute named `multigraph_input`. 

The suggestion of always treating the input as a multigraph representation would **not** allow edge attributes with a dict as value.  I'm not sure it would allow dict-of-dict-of-dict to be used as input to `MultiGraph` either. And that is probably much more common -- if you have a graph dodod representation and want to make a MultiGraph with it. But I don't actually know what is common or not common really. I'm just guessing.

dschult commented:
A more concise response would have been:
- we need to allow users to specify `multigraph_input=False` even for a MultiGraph constructor.
- it is true that this PR does not make #4824 work without their syntax having to add the multigraph_input keyword argument. But it would allow the user to get what they want via the `MultiGraph` constructor by adding such a keyword -- just as they do when using `to_networkx_graph`.

rossbar reviewed (commented):
Might this require some sort of modification to the docstring as well? It's tricky because it's still just the `attr` parameter, but now one of the keys has a special meaning. I'm not sure exactly what's the best way to call the user's attention to that. Should it be an explicit kwarg? I *think* this is sort of repeating @MridulS 's question from the discussion, but maybe I'm way off - please disregard this if the comment doesn't make sense.

dschult commented on networkx/classes/tests/test_multigraph.py:
```diff
@@ -182,6 +182,22 @@ def test_data_input(self):
 
+    def test_data_multigraph_input(self):
+        G = self.Graph({1: {2: {4: {}}}}, multigraph_input=True)
+        if G.is_directed():
+            expected = [(1, {2: {4: {}}}), (2, {})]
```
This test function is used to test `MultiDiGraph` as well as `MultiGraph`.  We could overwrite the method in `tests/test_multidigraph.py` if that is better.  I guess that is probably better style and perhaps less confusing. I'll also check how much of the method writing is written to handle both directed and undirected (and thus reduce overwriting methods)... Thanks for asking about that.

dschult commented:
Making this new option clear and easy to use by adding test to the the doc_strings is going to be challenging.
It requires going into the depth of dododod data structure. Where is the best place for that?  I'll try to find a place for it. The docs about this feature within `to_networkx_graph` do not describe it fully -- I'll look at that too.

MridulS commented:
These are all just opinions, feel free to ignore them :)

> we need to allow users to specify multigraph_input=False even for a MultiGraph constructor.

How about something like this?
```
diff --git a/networkx/classes/multigraph.py b/networkx/classes/multigraph.py
index 1779a616..c8078992 100644
--- a/networkx/classes/multigraph.py
+++ b/networkx/classes/multigraph.py
@@ -306,7 +306,10 @@ class MultiGraph(Graph):

         """
         self.edge_key_dict_factory = self.edge_key_dict_factory
-        Graph.__init__(self, incoming_graph_data, **attr)
+        if incoming_graph_data is not None and 'multigraph_input' not in attr:
+            Graph.__init__(self, incoming_graph_data, multigraph_input=True,  **attr)
+        else:
+            Graph.__init__(self, incoming_graph_data, **attr)
```
We can deal with both `multigraph_input` bool values and if the user wants they can explicitly turns it False.

``` python
In [1]: import networkx

In [2]: g = networkx.MultiGraph({'a': {'b': {0: {'w': 200}, 1: {'w': 201}}}})

In [3]: print(g.edges)
[('a', 'b', 0), ('a', 'b', 1)]

In [4]: g = networkx.MultiGraph({'a': {'b': {0: {'w': 200}, 1: {'w': 201}}}}, multigraph_input=False)

In [5]: print(g.edges)
[('a', 'b', 0)]
```

The reason I'm thinking of this change is to make this roundtrip work as default behaviour without the need to provide `multigraph_input` as I guess if we are creating a MultiGraph it should probably expect a `multigraph_input` by default? (I would probably expect this).
``` python
In [14]: h = networkx.MultiGraph()

In [15]: h.add_edges_from([('a', 'b', {"w": 200}), ('a', 'b', {"w": 201})])
Out[15]: [0, 1]

In [16]: g = networkx.MultiGraph(networkx.to_dict_of_dicts(h))

In [17]: networkx.is_isomorphic(g, h)
Out[17]: True
```

dschult commented:
Nice -- yes, I see now that the round-trip of `to_dict_of_dict` and `from_dict_of_dict` is probably good to establish for MultiGraphs.  As proposed, that will break dodod input to a MultiGraph because a dododod is expected when `multigraph_input is True`.  But could we use introspection of the `incoming_graph_data` to make sure it will work?  ... or maybe better, a try/except that catches the error if the input doesn't work when the default is applied.  

I'll try to put together a bunch of tests that define how we expect the graph constructors to treat input dict_of_dict data -- starting from the examples @MridulS provides above. I'll try to base them also after the tests for `to_networkx_graph`.

dschult commented:
Here are some tests....  Is this what you would expect to happen? 
(Note: `self.Graph` is the class we are testing. This is supposed to work for nx.MultiGraph and nx.MultiDiGraph)

I chose the default so that roundtrip works without adding a flag input, even though that changes current behavior.  

The second test function looks at the cases that are `dododod` or `dodod` or `dict-of-list` but for which we might not want or can't make `multigraph_input` work.

```python
    def test_data_multigraph_input(self):
        # standard case with edge keys and edge data
        edata0 = dict(w=200, s="foo")
        edata1 = dict(w=201, s="bar")
        keydict = {0: edata0, 1: edata1}
        dododod = {"a": {"b": keydict}}

        # possible desired results
        multiple_edge = [("a", "b", 0, edata0), ("a", "b", 1, edata1)]
        single_edge = [("a", "b", 0, keydict)]

        # test each form of multgraph_input
        cases = [ 
            (dododod, True, multiple_edge),
            (dododod, False, single_edge),
        ]   
        for dod, mgi, edges in cases:
            G = self.Graph(dod, multigraph_input=mgi)
            assert list(G.edges(keys=True, data=True)) == edges

        # test round-trip to_dict_of_dict and MultiGraph constructor
        G = self.Graph(dododod, multigraph_input=True)
        for mgi in [True, False]:
            H = self.Graph(nx.to_dict_of_dicts(G), multigraph_input=mgi)
            assert nx.is_isomorphic(G, H) == mgi 

    def test_data_non_multigraph_input(self):
        # other cases for dict incoming_graph_data
        edata = { 
            "traits": {"w": 200, "s": "foo"},
            "graphics": {"color": "blue", "shape": "box"},
        }   
        dodod1 = {"a": {"b": edata}}  # dicts as values in edgedata
        dodod2 = {"a": {"b": {"w": 200, "s": "foo"}}}  # no dicts as values in edgedata
        dodod3 = {"a": {"b": {"w": {"color": "blue"}, "s": "foo"}}}  # mix of both
        dol = {"a": ["b"]}  # dict-of neighbors iterable (called dict_of_lists in convert.py)

        # possible desirable results
        multiple_edge = [("a", "b", "traits"), ("a", "b", "graphics")]
        single_edge = [("a", "b", 0)]   # edge data will vary... but entire 3rd dict is the edge data 

        cases = [  # (dod, mgi, edges)
            (dodod1, True, multiple_edge),
            (dodod1, False, single_edge),
            (dodod2, True, single_edge),
            (dodod2, False, single_edge),
            (dodod3, True, single_edge),
            (dodod3, False, single_edge),
            (dol, True, single_edge),
            (dol, False, single_edge),
        ]   

        for dod, mgi, edges in cases:
            G = self.Graph(dod, multigraph_input=mgi)
            assert list(G.edges) == edges
            G = nx.to_networkx_graph(dod, create_using=self.Graph, multigraph_input=mgi)
            assert list(G.edges) == edges
```

dschult commented:
Looks like changing the default breaks a couple of places in our own codebase where it relies on `nx.MultiGraph` treating `dodod` input as `multigraph_input=False`.  I'll figure out a way to handle these cases... (they cause an exception that is currently handled by assuming they are `dict-of-list`.

dschult commented:
Because of the following change in API, I have some questions: 
- the default value of `multigraph_input` used in `nx.MultiGraph` and `nx.MultiDiGraph` used to effectively be `False`.  We are now changing it to `True`.  

Questions:
- Should this require a deprecation? It changes the API, but it is a feature that was hardly used and can be seen as fixing a defect instead of changing the API.
- Notice that this will make the default different from the default for `to_networkx_graph`. That breaks the mirrored behavior of `H = nx.MultiGraph(dod)` and `H = nx.to_networkx_graph(dod)`.   Should we worry about these giving different results? That is, should we also change the default input for `nx.to_networkx_graph`?

One reason for the current default is that "most" dict inputs are either dict-of-list, or dict-of-dict-of-dict without edge keys.  In the special case of `nx.MultiGraph` we know we are dealing with a multigraph. So that changes the likelihood of a dict-of-dict-of-dict-of-dict with edge keys.  It might be tempting to make the default value depend on whether `create_using` is a multigraph, but even our own tests show that it is common to want to initialize a multigraph using a  d-o-d-o-d.  

My current view is to make the default for `nx.MultiGraph(dod)` use `multigraph_input=True` while `nx.to_networkx_graph(dod, create_using=nx.MultiGraph)` defaults to `multigraph_input=False`.  But I seem to change preference from one day to the next.  Thoughts?

dschult commented:
Bottom line summary of what is in this PR:
1. Added code to MultiGraph.__init__  (and for MultiDiGraph) that 
    a. adds a multigraph_input parameter that only is used when the `incoming_graph_data` parameter is a dict.
    b. does minor preprocessing of the multigraph_input flag (True/False/None).
        i. if True, try `from_dict_of_dict` processing with multigraph turned on and if it doesn't work, raise an error.
        ii. if False, use the process used by `to_networkx_graph` where we ignore edge keys and load edges as dict-of-dict or as dict-of-list, whichever works.
        iii. if None:  try `from_dict_of_dict` processing with multigraph turned on and if it doesn't work, try again with multigraph processing turned off and if it doesn't work try dict-of-list.
2. Tweaked convert.py to make `to_networkx_graph` raise an error if multigraph_input is True and the input is not a complete and valid multigraph representation.  I also updated the tests for convert -- and found a comment that mentions that the current code can't process multigraph_input well because it can't tell whether to add both edges or not.  Well -- this fixes that problem. So the behavior of one current test has changed -- and apparently for the better.
3. Added a bunch of multidigraph input processing tests to multigraph and multidigraph.

I hope we now have an interface where the default for any dict input to multi(di)graph is to try loading it as a full implementation of multiedges with keys, if that doesn't work, then try again without the keys but with edge data and if that doesn't work then try without keys and without data (dict-of-iterable).  This changes the behavior of `to_networkx_graph` when `multigraph_input is True` because it now raises an error if the data is not fully multigraph. But if the default is used, it handles all the different kinds like it used to.  So, I believe no deprecation is needed.

MridulS commented:
This looks great to me! thanks @dschult!

> Should this require a deprecation? It changes the API, but it is a feature that was hardly used and can be seen as fixing a defect instead of changing the API.

I see this as a bug fix too, so this can just show up in the API changes part of the release log.

> Notice that this will make the default different from the default for to_networkx_graph. That breaks the mirrored behavior of `H = nx.MultiGraph(dod)` and `H = nx.to_networkx_graph(dod)`. Should we worry about these giving different results? That is, should we also change the default input for nx.to_networkx_graph?

IMO `H = nx.MultiGraph(dod)` and `H = nx.to_networkx_graph(dod)` are independent things, for the first one the user creates a multigraph explicitly so they know what they are expecting, in the other case they are asking networkx to infer the graph object for them, so I think it's okay for these functions to not be mirrored. 


> My current view is to make the default for `nx.MultiGraph(dod)` use `multigraph_input=True` while `nx.to_networkx_graph(dod, create_using=nx.MultiGraph)` defaults to `multigraph_input=False`. But I seem to change preference from one day to the next. Thoughts?

I think we should default to `multigraph_input=True` with `nx.to_networkx_graph(dod, create_using=nx.MultiGraph)` too, as the user could see that equivalent to `nx.MultiGraph(dod)`.

IMO it's okay if `nx.MultiGraph(dod)` != `nx.to_networkx_graph(dod)` but `nx.MultiGraph(dod)` == `nx.to_networkx_graph(dod,  create_using=nx.MultiGraph)`.

Let me know if this makes sense :)

dschult commented:
Thanks @MridulS for these comments.

The problem with enforcing `multigraph_input=True` for the cases when we are creating a multigraph is that often (including many places in our own tests) we want the `incoming_graph_data` to **not** be a multigraph.

And that makes sense -- we use the constructors to convert from one type to another type (though that isn't using d-o--d data structure.  For example: `MG = nx.MultiGraph(G)`.
And we also use d-o-d structure to construct a graph that we want to be a MultiGraph even though the input is not.
```python
dol = {1: [2], 2: [3]}
MG = nx.MultiGraph(dol)

or

dod = {0: {1: {'weight': 1.5}}, ....}
MG = nx.MultiGraph(dod)
```

Really, the type of the incoming data object should be indepedent of the type of the constructed graph.
I think that is true for both `to_networkx_graph(...)` and for `nx.MultiGraph(...)`.  Thoughts?

dschult commented:
I agree --- the round-trip should work by default without any special handling by the user.

OK... So I think this is ready then...

rossbar commented on networkx/classes/multidigraph.py:
```diff
@@ -277,6 +278,19 @@ def __init__(self, incoming_graph_data=None, **attr):
             packages are installed the data can also be a NumPy matrix
             or 2d ndarray, a SciPy sparse matrix, or a PyGraphviz graph.
 
+        multigraph_input : bool or None (default None)
+           Note: Only used when `incoming_graph_data` is a dict.
+           If True, `incoming_graph_data` is assumed to be a
+           dict-of-dict-of-dict-of-dict structure keyed by
+           node to neighbor to edge keys to edge data for multi-edges.
+           A NetworkXError is raised if this is not the case.
+           If False, :func:`to_networkx_graph` is used to try to determine
+           the dict's graph data structure as either a dict-of-dict-of-dict
+           keyed by node to neighbor to edge data, or a dict-of-iterable
+           keyed by node to neighbors.
+           If None, the treatment for True is tried, but if it fails,
+           the treatment for False is tried.
+
```
It looks like the [rendered docstrings](https://167-2055378-gh.circle-artifacts.com/0/doc/build/html/reference/classes/multigraph.html) are being built from the class docstring rather than the `__init__` docstring, so I think this should be copied up to the `Parameters` section in the class docstring as well. Same goes for the DiGraph

dschult replied:
Thank you!!  I missed that completely.  Looks like I mis-indented the description too, so I fixed that.
If you get a chance to see whether this did what you suggested, great -- if not, enjoy your gatherings! and good luck with your trip!

rossbar replied:
circleci still seems to be having problems, but it looks good now on a local build! :+1:

rossbar reviewed (approved):
LGTM - I think the interface is well-documented and this avoids special-casing any keys in `attr`, which is good!

The only thing is that I think the parameter description should be copied from the `__init__` docstring up into the Parameters section of the class docstring itself, so that it is more visible in the html docs.
